feat: add OpenLineage EventListener plugin

mgorsk1 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Adds OpenLineage plugin that uses EventListener SPI to transform and send lineage information to OpenLineage compatible API.

co-authored with @takezoe 

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# General
* Add [OpenLineage](https://openlineage.io/) event listener plugin. ({issue}`21265`)
```

dechoma commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageClient.java:
```diff
@@ -0,0 +1,196 @@
+    private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+    private final String url;
+    private static final ObjectMapper objectMapper = createMapper();
+    private Integer retryCount;
+    private Duration retryDelay;
+    private Duration maxDelay;
+    private double backoffBase;
```
shouldn't those variable be `final` ? 
they are initialized once in the constructor and do not seem to change

mgorsk1 replied:
true

dechoma commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageClient.java:
```diff
@@ -0,0 +1,196 @@
+
+public class OpenLineageClient
+{
+    private final HttpClient client = new JettyHttpClient();
```
maybe it would be more convenient in terms of tests to inject JettyHttpClient to the OpenLineageClient's constructor instead of hardcoding it here

dechoma commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageClientConfig.java:
```diff
@@ -0,0 +1,114 @@
+{
+    private String url;
+    private Optional<String> apiKey = Optional.empty();
+    private int retryCount;
```
it defaults to 0, shouldn't be there some retries by default ?

mgorsk1 replied:
ok, will make it 3

dechoma commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,303 @@
+        if (this.queryStatisticsFacetEnabled) {
+            OpenLineage.RunFacet trinoQueryStatisticsFacet = ol.newRunFacet();
+
+            for (Field field : queryStatistics.getClass().getDeclaredFields()) {
+                field.setAccessible(true);
+                trinoQueryStatisticsFacet
+                        .getAdditionalProperties()
+                        .put(field.getName(), String.valueOf(field.get(queryStatistics)));
+            }
```
I wonder if using here manual mapping of properties instead of reflection wouldn't be beneficiary in terms of stability and maintenance, currently we are not sure what should be expected in statistics facet

mgorsk1 replied:
I assumed additional facets do not need to have such strict contracts and this is much more convenient to maintain. Let's see if someone else has opinion in either direction.

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageClient.java:
```diff
@@ -0,0 +1,196 @@
+            throws JsonProcessingException
+    {
+        String json = objectMapper.writeValueAsString(runEvent);
+        logger.info(json);
```
true

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageClient.java:
```diff
@@ -0,0 +1,196 @@
+public class OpenLineageClient
+{
+    private final HttpClient client = new JettyHttpClient();
+    private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
```
added

alprusty commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,312 @@
+        UUID runID = getQueryId(queryCompletedEvent.getMetadata());
+
+        try {
+            OpenLineage.RunEvent event = getCompletedEvent(runID, queryCompletedEvent);
```
Shouldn't we just capture lineage for mutation queries i.e. DDLs, DMLs ?
Without [QueryType](https://github.com/trinodb/trino/blob/master/core/trino-spi/src/main/java/io/trino/spi/resourcegroups/QueryType.java#L21-L23) filtering, it is also going to emit lineage events for `SELECT` queries as well. Trino being a shared query engine for adhoc query analysis and dashboarding use cases, will normally see a huge volume of SELECT queries compared to DMLs or DDLs. So in the lineage backend system(e.g. Marquez) we will see a lot of lineage nodes for selects.


e.g. below set of queries will contribute 5 lineage events but from dataset lineage point of view we are interested to see the how `test_table` and `test_table2` got created from `nation`.
```
create table memory.default.test_table as select * from tpch.sf1.nation;
create table memory.default.test_table2 as select * from tpch.sf1.nation;
select * from from tpch.sf1.nation;
select * from from tpch.sf1.nation;
select * from from tpch.sf1.nation;
```


Dataset owners would rather prefer to see which ddl/dml queries are mutating their datasets and would be interested to see respective upstream and downstream lineage events, as compared no-transformation vanilla select queries.

`queryCompletedEvent.getContext().getQueryType()` will get you the `QueryType` here.

We can also think about giving this option back to user, so they can dictate if they are interested in seeing lineage for all query types or just a subset. That configuration can go to `OpenLineageListenerConfig`

mgorsk1 replied:
very good point!

mgorsk1 replied:
addressed here https://github.com/mgorsk1/trino/pull/1

mobuchowski replied:
I think this is something that could be configurable by plugin user - not sure how Trino plugins do it - but getting events for `SELECT` queries is a feature that some people would like.

mgorsk1 replied:
yep, I made it into a config option with sensible defaults

alprusty commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,312 @@
+        trinoMetadata.ifPresent(runFacet -> runFacetsBuilder.put("trino.metadata", runFacet));
+        trinoQueryStatistics.ifPresent(runFacet -> runFacetsBuilder.put("trino.queryStatistics", runFacet));
+        trinoQueryContext.ifPresent(runFacet -> runFacetsBuilder.put("trino.queryContext", runFacet));
+
```
Instead of emitting two lineage nodes, one for `START` and one for `COMPLETE`, can we not just rely on `QueryCompletedEvent` and send just one `COMPLETE` event if query execution was successful.

Also to expand my other follow up above around mutation queries, If It is a `CTAS` query that failed, do we really need to emit a `START` event and `FAIL` event or just skip sending any event because query has failed.

mgorsk1 replied:
in my opinion we shouldn't leave out START events as:
- state change START > COMPLETE can be a source of valuable information
- having both states/events leaves more audit information

@mobuchowski do you have an opinion about just leaving COMPLETE events? are there any OL integrations leaving START events on purpose?

mobuchowski replied:
>are there any OL integrations where it makes sense to leave out START events on purpose?

Not on purpose. If it's possible to get `START` events - and I see it is - it would definitely be preferable for OpenLineage backends - @mgorsk1 is right.

>If It is a CTAS query that failed, do we really need to emit a START event and FAIL event or just skip sending any lineage event because query has failed.

OpenLineage backends can also take advantage of FAIL events, it's the same story - additional information.

alprusty commented:
cc @mosabua 
Could you please help here in reviews. 🙏

mgorsk1 commented:
also cc @wendigo

ebyhr commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/OpenLineageListenerTest.java:
```diff
@@ -0,0 +1,209 @@
+@TestInstance(PER_METHOD)
+class TestHttpEventListener
+{
+    private MockWebServer server;
```
We avoid using mock in tests. https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-mocks

We should use docker image instead.

mgorsk1 replied:
refactored tests so it doesn't require mocking apis

ebyhr reviewed (commented):
Please fix the commit title to `Add OpenLineage plugin`. Also, I would recommend including `@takezoe` (original author in https://github.com/takezoe/trino-openlineage) as the co-author in the commit.

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+  - None.
+
+* - openlineage-event-listener.connect-retry-count
+  - The number of retries on server error. A server is considered to be
```
```suggestion
  - The number of retries on server error. A server is considered
```

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+
+* - openlineage-event-listener.connect-retry-count
+  - The number of retries on server error. A server is considered to be
+    in an error state when the response code is 500 or higher
```
```suggestion
    in an error state when the response code is 500 or higher.
```

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+  - `3`
+
+* - openlineage-event-listener.connect-retry-delay
+  - Duration for which to delay between attempts to send a request
```
```suggestion
  - [Duration](prop-type-duration) between attempts to send a request.
```

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+    The formula used to calculate the delay is
+    `attemptDelay = retryDelay * backoffBase^{attemptCount}`.
+    Attempt count starts from 0. Leave this empty or set to 1 to disable
+    exponential backoff and keep constant delays
```
```suggestion
    exponential backoff and keep constant delays.
```

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+  - `2`
+
+* - openlineage-event-listener.connect-max-delay
+  - The upper bound of a delay between 2 retries. This should be
```
```suggestion
  - The upper bound of a delay between two retries. This should be
```

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+  - `1m`
+
+* - openlineage-event-listener.trino-host
+  - Trino hostname. Will be used to render Job Namespace.
```
```suggestion
  - Trino hostname. Used to render Job Namespace in OpenLineage.
```

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+
+* - openlineage-event-listener.trino-host
+  - Trino hostname. Will be used to render Job Namespace.
+  - `localhost`
```
I think it should NOT have a default value. And certainly not the generic term `localhost`.

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+  - `localhost`
+
+* - openlineage-event-listener.trino-port
+  - Trino port. Will be used to render Job Namespace.
```
```suggestion
  - Trino port. Used to render Job Namespace in OpenLineage.
```

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+  - `8080`
+
+* - openlineage-event-listener.facets-metadata-enabled
+  - Should Trino Metadata facet be included into run facet.
```
explain more what info that is or link to info elsewhere

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+  - `true`
+
+* - openlineage-event-listener.facets-query-context-enabled
+  - Should Query Context facet be included into run facet.
```
explain more what info that is or link to info elsewhere

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+  - `true`
+
+* - openlineage-event-listener.facets-query-statistics-enabled
+  - Should Query Statistics facet be included into run facet.
```
explain more what info that is or link to info elsewhere

mosabua commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,133 @@
+  - `true`
+
+* - openlineage-event-listener.include-query-types
+  - Which types of queries should be taken into account when emitting lineage information. List of values split by comma. Each value must be matching `io.trino.spi.resourcegroups.QueryType` enum. Query types not included here will be filtered out. 
```
```suggestion
  - Which types of queries should be taken into account when emitting lineage information. List of values split by comma. Each value must be matching `io.trino.spi.resourcegroups.QueryType` enum. Query types not included here are filtered out. 
```

mosabua reviewed (changes requested):
First review on docs done. Some of it has impact on code.

mgorsk1 commented:
Thanks [mosabua](https://github.com/mosabua) for thorough review, I've made corrections according to your suggestions.

mobuchowski commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageClient.java:
```diff
@@ -0,0 +1,202 @@
+import static io.airlift.http.client.Request.Builder.preparePost;
+import static io.airlift.http.client.StatusResponseHandler.createStatusResponseHandler;
+
+public class OpenLineageClient
```
Why not use `OpenLineageClient` from `openlineage-java` directly, but reimplement it?

mgorsk1 replied:
good point, i didn't want to dramatically diverge from original implementation but indeed ol client would be preferred. let me refactor this, it will be easier to extend with different transport types

mgorsk1 replied:
done

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -212,19 +252,18 @@ private OpenLineage.RunEvent getStartEvent(UUID runID, QueryCreatedEvent queryCr
     }
 
-    private OpenLineage.RunEvent getCompletedEvent(UUID runID, QueryCompletedEvent queryCompletedEvent)
+    public OpenLineage.RunEvent getCompletedEvent(UUID runID, QueryCompletedEvent queryCompletedEvent)
```
there is no `QueryEvent` afaik and sadly `QueryCreatedEvent` and `QueryCompletedEvent` do not implement same interface (there is no interface at all) so it's not possible to generalize this

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -43,34 +50,77 @@
+                            new HttpConfig(
+                                    new URI(config.getUrl()),
+                                    config.getEndpoint(),
+                                    config.getTimeout(),
```
it's inherited from OL client so I am not sure - is there a reason for both in the client @mobuchowski ?

mobuchowski replied:
Use `getTimeoutInMillis` - it's preferred approach now. Two different units are there due to bug https://github.com/OpenLineage/OpenLineage/issues/2474

alprusty commented on plugin/trino-openlineage/pom.xml:
```diff
@@ -0,0 +1,161 @@
+        <version>445-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+
```
Most of the `artifactId` are kept consistent to match the name of project directory inside `plugin`.

Should we change the artifactId to `trino-openlineage` instead ?

mgorsk1 replied:
sure!

alprusty commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,129 @@
+    {
+        return url;
+    }
+
```
Should we make it consistent to how Spark and Flink configs are given
https://openlineage.io/docs/client/java/configuration/#transports

e.g.
`trino.openlineage.transport.url`
`trino.openlineage.transport.endpoint`
`trino.openlineage.transport.auth.type`
`trino.openlineage.transport.auth.apiKey`

mgorsk1 replied:
I've adjusted naming convention, please see https://github.com/trinodb/trino/pull/21265/commits/a40786a4fedd9c739a22dd177bf4a4c4db14bbee

alprusty commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,370 @@
+    private List<OpenLineage.InputDataset> buildInputs(QueryIOMetadata ioMetadata)
+    {
+        return ioMetadata.getInputs().stream().map(inputMetadata ->
+                ol.newInputDatasetBuilder()
```
As per dataset naming conventions
https://openlineage.io/docs/spec/naming/#dataset-naming

Shouldn't 
`namespace` be set as `{host}:{port}`
and `name` as `{catalog}.{schema}.{table}` here ?

alprusty replied:
<img width="1055" alt="Screenshot 2024-04-15 at 7 44 38 AM" src="https://github.com/trinodb/trino/assets/110063541/841172e1-8413-4a02-9425-9d5624df173a">

mgorsk1 replied:
sure, will align it. although this PR preceds refactor of dataset naming, so I would rather finally merge this rather than wait for continuous changes in OL conventions :)

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageListenerConfig.java:
```diff
@@ -0,0 +1,110 @@
+    {
+        return transport;
+    }
+
```
I would stick to having this in config because server address might be changing (for example in kubernetes) and we don't get DNS name of trino coordinator anywhere in event metadata.

mgorsk1 commented:
@mosabua pinging for re-review 🙏

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,372 @@
+                                    config.getUrlParams(),
+                                    config.getHeaders())));
+        }
+        else if (listenerConfig.getTransport() == OpenLineageTransport.CONSOLE) {
+            logger.info("Using ConsoleTransport");
+            clientBuilder.transport(new ConsoleTransport());
+        }
```
We can extract some sort of an interface aka Decorator which would apply these changes on top `OpenLineageClient.Builder` and each authentication mechansim can have its own Decorator which would be binded as a part of its module.

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,372 @@
+            throw new Exception("Unsupported Transport Type");
+        }
+
+        String[] disabledFacets = listenerConfig.getDisabledFacets().stream().map(OpenLineageTrinoFacet::getText).toArray(String[]::new);
```
`.map` and subsequent lines chould be moved to next line,

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,129 @@
+        return this;
+    }
+
+    public Integer getTimeoutInMillis()
+    {
+        return timeoutInMillis;
+    }
+
+    @ConfigDescription("Timeout (milliseconds) when making HTTP Requests.")
+    @Config("openlineage-event-listener.openlineage.transport.timeoutInMillis")
+    public OpenLineageClientHttpTransportConfig setTimeoutInMillis(Integer timeoutInMillis)
+    {
+        this.timeoutInMillis = timeoutInMillis;
+        return this;
+    }
```
My intention was to stick 1:1 to OL client data types so other OL integrations have same configuration approach - for example in spark config is also explicitly what client expects https://openlineage.io/docs/integrations/spark/configuration/transport

Praveen2112 replied:
But in case of Trino we tend to make all the configuration (across plugin or various module) to be similar way so any timeout specific configuration property would be generally injected as `Duration` from Airlift's unit library which would be unified across the project.

mgorsk1 replied:
added via https://github.com/trinodb/trino/pull/21265/commits/a36ed5a509146e10397e62b86769b6038f5e40fe

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,376 @@
+        OpenLineage.RunFacet trinoQueryStatisticsFacet = openLineage.newRunFacet();
+
+        for (Field field : queryStatistics.getClass().getDeclaredFields()) {
+            field.setAccessible(true);
+            trinoQueryStatisticsFacet
+                    .getAdditionalProperties()
+                    .put(field.getName(), String.valueOf(field.get(queryStatistics)));
```
Why can't we store the Json representation of it directly ?

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,376 @@
+        runFacetsBuilder.put(OpenLineageTrinoFacet.TRINO_QUERY_CONTEXT.getText(),
+                getTrinoQueryContextFacet(queryCreatedEvent.getContext()));
+
+        OpenLineage.RunEvent startEvent =
```
Can we have return statement here ? i.e return openLineage.newRunEventBuilder()`

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,376 @@
+        runFacetsBuilder.put(OpenLineageTrinoFacet.TRINO_QUERY_STATISTICS.getText(),
+                getTrinoQueryStatisticsFacet(queryCompletedEvent.getStatistics()));
+
+        OpenLineage.RunEvent completedEvent =
```
Ditto

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,376 @@
+    {
+        return ioMetadata.getInputs().stream().map(inputMetadata ->
+                openLineage.newInputDatasetBuilder()
+                        .namespace(getDatasetNamespace())
```
Can we directly use `jobNamespace` here ?

mgorsk1 replied:
I've just aligned everything to `this.namespace`

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,376 @@
+                                    ).build()
+                            ).build());
+        }
+        else {
+            return ImmutableList.of();
+        }
```
else is redundant - we can directly `return ImmutableList#of`

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,376 @@
+
+    private List<OpenLineage.InputDataset> buildInputs(QueryIOMetadata ioMetadata)
+    {
+        return ioMetadata.getInputs().stream().map(inputMetadata ->
```
For determining can we use `queryCompletedEvent.getMetadata().getTables()` which provide additional information on whether is a table is referenced directly or via a View this could provide a better lineage information.

mgorsk1 replied:
I checked it and don't see much benefit over QueryIOMetadata. However, I could use your knowledge here - is there a way to get info about underlying system related to given table/view? For example, we can do `inputMetadata.getCatalogName()` but can we get some more information about this catalog? Like whether it's druid, hive, etc ?

Praveen2112 replied:
`queryCompletedEvent.getMetadata().getTables()#isDirectlyReferenced` provides an information if a table is referenced directly or is it referred via a view in this case we could point lineage only to the views instead of mapping them to the underlying table as well. Plus it also captures some additional information like viewText and referencedChain - So I guess for input details it makes much sense to use `queryCompletedEvent.getMetadata().getTables()`

Praveen2112 replied:
> Like whether it's druid, hive, etc ?

I don't think we have an option to fetch those information today - But it could be possible

alprusty commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageTransport.java:
```diff
@@ -0,0 +1,21 @@
+public enum OpenLineageTransport {
+    CONSOLE,
+    HTTP,
+    // @todo add kafka support
+    //KAFKA,
```
Trino issue has been created to add `KAFKA` transport support
https://github.com/trinodb/trino/issues/21599

I am working on a PR to add this support.
cc @mgorsk1 @Praveen2112

mgorsk1 replied:
removed the comment accordingly

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageListenerConfig.java:
```diff
@@ -0,0 +1,110 @@
+public class OpenLineageListenerConfig
+{
+    private OpenLineageTransport transport = OpenLineageTransport.CONSOLE;
+    private String trinoHost;
+    private Integer trinoPort;
```
OpenLineage enforces cluster/namespace convention so I wouldn't want to leave this open. https://openlineage.io/docs/spec/naming

mgorsk1 commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,241 @@
+(supported-transport-types)=
+### Supported Transport Types
+
+- `CONSOLE` - sends OpenLineage JSON event to Trino coordinator standard output.
```
yes https://github.com/OpenLineage/OpenLineage/blob/main/client/java/src/main/java/io/openlineage/client/transports/ConsoleTransport.java

mgorsk1 commented on docs/src/main/sphinx/admin/event-listeners-openlineage.md:
```diff
@@ -0,0 +1,241 @@
+    - Property name
+    - Description
+    - Default
+*
+    - openlineage-event-listener.openlineage.transport.url
+    - URL of OpenLineage . Required if `HTTP` transport is configured.
+    - None.
+*
+    - openlineage-event-listener.openlineage.transport.endpoint
+    - Custom path for OpenLineage compatible endpoint. If configured, there
+      cannot be any custom path within 
+      `openlineage-event-listener.openlineage.transport.url`.
+    - `/api/v1`.
```
these are separete openlineage http transport configuration options so I think it's better to configure them separately - otherwise i'd need to introduce code that splits this in some arbitrary way and passes relevant parts to HttpConfig instance

```
new HttpConfig(new URI(config.getUrl()), config.getEndpoint(), ...)
```

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,131 @@
+        return endpoint;
+    }
+
+    @ConfigDescription("Custom path . Explicitly set the scheme https:// to use symmetric encryption")
```
we don't. I think this is just information to the user - if you explicitly set scheme to https:// then symetric encryptrion will be used. we have same config description in http event listener for corresponding config option.

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,131 @@
+        return this;
+    }
+
+    public Optional<String> getApiKey()
```
if I add @NotEmpty here then it will fail if the value is not provided at all, defeating the purpose of Optional

mgorsk1 commented:
@Praveen2112 I've addressed your remarks including adding integration tests: https://github.com/trinodb/trino/pull/21265/commits/bdca31ecf2d42aaed83ca228844053cc17bd2cbd

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,131 @@
+        return url;
+    }
+
+    @ConfigDescription("URL of receiving server. Explicitly set the scheme https:// to use symmetric encryption")
```
OL client doesn't have configuration for keystore/trustore afaik, when it does then yes we would also need to extend this.

ebyhr commented on plugin/trino-openlineage/pom.xml:
```diff
@@ -0,0 +1,215 @@
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>trino-openlineage</artifactId>
```
Add this module to `core/trino-server/src/main/provisio/trino.xml`

mgorsk1 replied:
done

ebyhr commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/FacetNotSupportedException.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.Arrays;
+
+public class FacetNotSupportedException
```
Remove this class and use IllegalArgumentException instead.

mgorsk1 replied:
done

ebyhr commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,387 @@
+            }
+        }
+        else {
+            logger.debug("Query type "
+                    + queryCompletedEvent.getContext().getQueryType().toString()
+                    + " not supported. Supported query types: "
+                    + this.includeQueryTypes);
```
Avoid string concat in logger arguments.

ebyhr commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,387 @@
+
+    private String getDatasetName(String catalogName, String schemaName, String tableName)
+    {
+        return String.format("%s.%s.%s", catalogName, schemaName, tableName);
```
Static import `format` method. 
https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#prefer-string-formatting

ebyhr commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageListenerConfig.java:
```diff
@@ -0,0 +1,112 @@
+    private URI trinoURI;
+    private List<OpenLineageTrinoFacet> disabledFacets = ImmutableList.of();
+
+    private List<QueryType> includeQueryTypes = Arrays.asList(QueryType.DELETE,
+            QueryType.INSERT, QueryType.MERGE, QueryType.UPDATE, QueryType.ALTER_TABLE_EXECUTE);
```
Use ImmutableList builder. Also, please consider changing to Set. 
```java
    private List<QueryType> includeQueryTypes = ImmutableList.<QueryType>builder()
            .add(QueryType.DELETE)
...
            .build();
```

ebyhr commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/MarquezIntegrationTest.java:
```diff
@@ -0,0 +1,121 @@
+        server = closeAfterClass(new MarquezServer("0.46.0"));
+        marquezURI = server.getMarquezUri().toString();
+
+        Map<String, String> config = ImmutableMap.of(
+                "openlineage-event-listener.openlineage.transport.type", "HTTP",
+                "openlineage-event-listener.openlineage.transport.url", server.getMarquezUri().toString(),
+                "openlineage-event-listener.trino.uri", trinoURI);
```
nit: Inline and use builder.

mgorsk1 replied:
done

ebyhr commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/OpenLineageListenerQueryRunner.java:
```diff
@@ -0,0 +1,69 @@
+    public static QueryRunner createOpenLineageRunner(Map<String, String> listenerProperties)
+            throws Exception
+    {
+        QueryRunner queryRunner = null;
```
There's no need to initialize separately. 
```java
        QueryRunner queryRunner = DistributedQueryRunner.builder(createSession())
                .setEventListener(new OpenLineageListenerFactory().create(listenerProperties))
                .build();
```

mgorsk1 replied:
but then I won't be able to pass it to closeAllSuppress upon some error no?

ebyhr commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/OpenLineageListenerQueryRunner.java:
```diff
@@ -0,0 +1,69 @@
+    {
+        QueryRunner queryRunner = null;
+        try {
+            Session s = createSession();
```
Please avoid abbreviations: https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-abbreviations

Also, we can inline this variable.

mgorsk1 replied:
sure

ebyhr commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/MarquezServer.java:
```diff
@@ -0,0 +1,120 @@
+    private static final int POSTGRES_PORT;
+    private static final String POSTGRES_USER;
+    private static final String POSTGRES_PASSWORD;
+    private static final String LATEST_VERSION;
```
Remove unused constant.

mgorsk1 replied:
done

ebyhr commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/MarquezServer.java:
```diff
@@ -0,0 +1,120 @@
+    private final GenericContainer<?> dockerContainerAPI;
+    private final GenericContainer<?> dockerContainerPostgres;
+
+    public MarquezServer()
+    {
+        this(LATEST_VERSION);
+    }
```
Change the default version to `0.46.0` instead.

mgorsk1 replied:
done

ebyhr commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/MarquezServer.java:
```diff
@@ -0,0 +1,120 @@
+
+        this.dockerContainerPostgres.start();
+
+        this.dockerContainerAPI = new GenericContainer<>("marquezproject/marquez:" + version)
```
Does Marquez and Postgres expose static port?

mgorsk1 replied:
Marquez does, I've changed the setup to use them (5000/5001 instead of 7000/7001)

ebyhr replied:
We should expose random port numbers. Please take a look at other test classes using testcontainers.

mgorsk1 replied:
they are exposed randomly:

```
CONTAINER ID   IMAGE                                                       COMMAND                  CREATED         STATUS                          PORTS                                                                                      NAMES
f58d41f6851c   marquezproject/marquez:0.46.0                               "/usr/src/app/entryp…"   4 seconds ago   Up 2 seconds                    0.0.0.0:33233->5000/tcp, :::33233->5000/tcp, 0.0.0.0:33232->5001/tcp, :::33232->5001/tcp   sleepy_lewin
7c2dc7bf16f5   postgres:14                                                 "docker-entrypoint.s…"   5 seconds ago   Up 4 seconds                    0.0.0.0:33231->5432/tcp, :::33231->5432/tcp                                                objective_dhawan
```

ebyhr commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/OpenLineageListenerQueryRunner.java:
```diff
@@ -0,0 +1,69 @@
+                .setSchema(SCHEMA)
+                .build();
+    }
+}
```
Could you add main method so that we can try this plugin easily?

mgorsk1 replied:
done

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    @ConfigDescription("API Key to use when authenticating against OpenLineage API")
+    @Config("openlineage-event-listener.openlineage.transport.apiKey")
```
the intention was to keep it consistent with regular http event listener but if you think it's better not to I can remove camelcase

ebyhr replied:
I understand your intention. However, please avoid camel case for consistency with any other plugins.

Praveen2112 commented on plugin/trino-openlineage/pom.xml:
```diff
@@ -0,0 +1,210 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.datatype</groupId>
+            <artifactId>jackson-datatype-jdk8</artifactId>
+            <version>2.17.0</version>
```
Any specific to use this pinned version ? We could use the version from referred from airlift right ?

mgorsk1 replied:
agree, it's fine not to pin it

Praveen2112 commented on plugin/trino-openlineage/pom.xml:
```diff
@@ -0,0 +1,210 @@
+        <dependency>
+            <groupId>io.openlineage</groupId>
+            <artifactId>openlineage-java</artifactId>
+            <version>1.12.0</version>
```
Can we move them to properties so that we could change it there for the next version updates

mgorsk1 replied:
of course

Praveen2112 commented on plugin/trino-openlineage/pom.xml:
```diff
@@ -0,0 +1,210 @@
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-classic</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-core</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Why do we need this dependency ?

mgorsk1 replied:
probably some leftovers, removed it

Praveen2112 commented on plugin/trino-openlineage/pom.xml:
```diff
@@ -0,0 +1,210 @@
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>com.squareup.okhttp3</groupId>
+            <artifactId>mockwebserver</artifactId>
+            <scope>test</scope>
+        </dependency>
```
If we could use a marquez image for integration test, do we still need to use `mockwebserver` ?

mgorsk1 replied:
no, will remove it

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,376 @@
+        @Override
+        public String getToken()
+        {
+            return this.token;
```
Any updates here ?

mgorsk1 replied:
sorry, missed this before. indeed, great catch - will do format("Bearer %s", this.token)

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,139 @@
+        return this;
+    }
+
+    public String getEndpoint()
+    {
+        return endpoint;
+    }
+
+    @Config("openlineage-event-listener.transport.endpoint")
+    @ConfigDescription("Custom path for API receiving the events.")
+    public OpenLineageClientHttpTransportConfig setEndpoint(String endpoint)
+    {
+        this.endpoint = endpoint;
+        return this;
```
Can we combine the url and endpoint config ?

mgorsk1 replied:
I configured them separately as this is the approach in OpenLineage client used in different integrations. Aditionally, having them separate unburdens users from having to know what is the default endpoint for OL lineage API.

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,376 @@
+        return UUID.nameUUIDFromBytes(queryMetadata.getQueryId().getBytes(StandardCharsets.UTF_8));
+    }
+
+    private OpenLineage.RunFacet getTrinoQueryContextFacet(QueryContext queryContext)
```
done for everything

Praveen2112 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/MarquezServer.java:
```diff
@@ -0,0 +1,122 @@
+    public MarquezServer(String version)
+    {
+        network = Network.newNetwork();
+        this.dockerContainerPostgres = new GenericContainer<>("postgres:14")
+                .withNetwork(network)
+                .withNetworkAliases(POSTGRES_HOST)
+                .withEnv("POSTGRES_USER", POSTGRES_USER)
+                .withEnv("POSTGRES_PASSWORD", POSTGRES_PASSWORD)
+                .withEnv("POSTGRES_DB", POSTGRES_DB)
+                .withStartupTimeout(Duration.ofSeconds(360));
```
Can we use `PostgreSQLContainer` here ?

mgorsk1 replied:
we can, will do

Praveen2112 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/MarquezServer.java:
```diff
@@ -0,0 +1,122 @@
+    @Override
+    public void close()
+    {
+        dockerContainerPostgres.close();
+        dockerContainerAPI.close();
+        network.close();
```
We could use `Closer` , register them closing logic for the container in the container and we could use `closer.close()` here

mgorsk1 replied:
great idea, done

Praveen2112 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/MarquezServer.java:
```diff
@@ -0,0 +1,122 @@
+                .withCopyFileToContainer(
+                        MountableFile.forClasspathResource("marquez.yaml"),
+                        MARQUEZ_CONFIG)
+                .withCommand("/opt/marquez/wait-for-it.sh", format("%s:%s", POSTGRES_HOST, POSTGRES_PORT), "--", "./entrypoint.sh")
```
Instead can we `dependsOn` which ensures this container would be started only after postgres

mgorsk1 replied:
done!

Praveen2112 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/MarquezServer.java:
```diff
@@ -0,0 +1,122 @@
+        this.dockerContainerAPI.start();
+    }
+
+    public String getPostgresUri()
```
This could be private

Praveen2112 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/TestMarquezIntegration.java:
```diff
@@ -0,0 +1,117 @@
+                .toString();
+
+        // Wait for marquez to process event
+        Thread.sleep(5000);
```
We could wrap the assertion via `assertEventually` instead of `Thread#sleep`

mgorsk1 replied:
great point

Praveen2112 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/TestMarquezIntegration.java:
```diff
@@ -0,0 +1,117 @@
+
+        URI trino = new URI(trinoURI);
+
+        HttpClient client = HttpClient.newHttpClient();
```
We could re-use the client across multiple tests right ?

mgorsk1 replied:
yes, moved it into private static final class attribute

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,376 @@
+        }
+    }
+
+    private OpenLineageClient getClient(OpenLineageListenerConfig listenerConfig, Optional<OpenLineageClientHttpTransportConfig> httpTransportConfig)
```
done

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,376 @@
+    {
+        OpenLineageClient.Builder clientBuilder = OpenLineageClient.builder();
+
+        if (listenerConfig.getTransport().equals(OpenLineageTransport.HTTP) && httpTransportConfig.isPresent()) {
```
I propose we do this upon PR introducing another transport type, wdyt?

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,340 @@
+    {
+        RunFacet queryContextFacet = openLineage.newRunFacet();
+
+        queryContextFacet
+                .getAdditionalProperties()
+                .put("serverAddress", queryContext.getServerAddress());
```
Is it possible to extract them to a method so that we could pass each individual argument or pass a map of additional properties. WDYT ?

mgorsk1 replied:
I don't fully get the question, can you rephrase please?

mgorsk1 replied:
how about 

```
    private RunFacet getTrinoQueryContextFacet(QueryContext queryContext)
    {
        RunFacet queryContextFacet = openLineage.newRunFacet();

        java.util.Map<String, String> properties = Maps.<String, String>newHashMap();

        properties.put("server_address", queryContext.getServerAddress());
        properties.put("environment", queryContext.getEnvironment());

        queryContext.getQueryType().ifPresent(queryType ->
                properties.put("query_type", queryType.toString()));

        queryContextFacet
                .getAdditionalProperties()
                .putAll(properties);

        return queryContextFacet;
    }
```

and same for all 3 methods creating facets?

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,340 @@
+
+    private List<InputDataset> buildInputs(QueryIOMetadata ioMetadata)
+    {
+        return ioMetadata.getInputs().stream().map(inputMetadata ->
```
To re-iterate on the same question - using `QueryIOMetadata#getInputs` doesn't recognize views and considers only normal table - so for a `CTAS` query `CREATE TABLE new_table AS SELECT * FROM view_1` where SQL under represented by view_1 would be `SELECT * FROM table_1` and the lineage would point to from`new_table -> table_1` and view information is potentially lost in this operation - This is not a recommended operation right ? For those types of query we need to map the the lineage between `new_table` and `view_1`

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,340 @@
+    private final String namespace;
+    private final Set<QueryType> includeQueryTypes;
+    private final ObjectMapper queryStatisticsMapper;
+    private static final Logger logger = Logger.get(OpenLineageListener.class);
```
Can we move `logger` to LNo: 59 and new line post that.

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,340 @@
+
+    @Inject
+    public OpenLineageListener(OpenLineageClient client, OpenLineageListenerConfig listenerConfig)
+            throws Exception
```
It doesn't throw this Exception right ?

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,340 @@
+            throws Exception
+    {
+        this.client = client;
+        this.namespace = listenerConfig.getTrinoURI().toString().replace(listenerConfig.getTrinoURI().getScheme(), "trino");
```
Instead of replacing the schema, can we enforce it to have `trino` as a part of schema ?

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,340 @@
+        this.client = client;
+        this.namespace = listenerConfig.getTrinoURI().toString().replace(listenerConfig.getTrinoURI().getScheme(), "trino");
+
+        this.includeQueryTypes = listenerConfig.getIncludeQueryTypes();
```
nit: `Immutable#copyOf`

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,340 @@
+    }
+
+    public RunEvent getStartEvent(UUID runID, QueryCreatedEvent queryCreatedEvent)
+            throws IllegalAccessException
```
Does it throws `IllegalAccessException` ?

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,340 @@
+        if (queryTypeSupported(queryCompletedEvent.getContext())) {
+            UUID runID = getQueryId(queryCompletedEvent.getMetadata());
+
+            try {
+                RunEvent event = getCompletedEvent(runID, queryCompletedEvent);
+                client.emit(event);
+            }
+            catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
```
Do we need this try-block, we don't see `getCompletedEvent` throws any `IllegalAccessException`

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListenerFactory.java:
```diff
@@ -0,0 +1,45 @@
+    public EventListener create(Map<String, String> config)
+    {
+        Bootstrap app = new Bootstrap(
+                        new OpenLineageListenerModule());
```
We could move it to previous line

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListenerModule.java:
```diff
@@ -0,0 +1,110 @@
+
+        public ApiKeyTokenProvider(String token)
+        {
+            this.token = token;
```
nit: rnn check here

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListenerModule.java:
```diff
@@ -0,0 +1,110 @@
+                OpenLineageListenerConfig.class,
+                config -> config.getTransport()
+                        .equals(OpenLineageTransport.HTTP),
+                conditionalBinder -> configBinder(conditionalBinder).bindConfig(OpenLineageClientHttpTransportConfig.class)));
```
We could create a new module which could capture the changes related to `HTTP` transport.

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListenerModule.java:
```diff
@@ -0,0 +1,110 @@
+                                    config.getHeaders())));
+        }
+        else if (listenerConfig.getTransport() == OpenLineageTransport.CONSOLE) {
+            clientBuilder.transport(new ConsoleTransport());
```
It is possible for us to introduce some sort of a interface like `MongoClientSettingConfigurator` we have in MongoDB which could use 
```
public interface OpenLineageClientConfigurator
{
    void configure( OpenLineageClient.Builder builder);
}
```

The logic related to each transport layer could be specific to each module.

mgorsk1 replied:
I think as we have just one configurable transport for now I would leave it as is - with introducing subsequent ones this of course can be generalized.

Praveen2112 replied:
It is technically two right `CONSOLE` and `HTTP` as far as new one - I guess Kafka - the implementation and testing could be a bit tricky and I think handling this prep work as a part of that PR would make it a bit complex. Feel free to ping me in slack if you need any help on abstracting it out.

mgorsk1 replied:
ok
- added `OpenLineageTransportBuilder` interface
- implemented above for console and http
- added conditional module installation in `OpenLineageListenerModule`

I think it's now quite generic and clear how to extend with new transport types

Praveen2112 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/TestMarquezIntegration.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Test
+    void testCreateTableAsSelectFrom()
```
We need some additional tests like creating table from view/stale materialized view/fresh MV and insert for similar operation.

mgorsk1 replied:
as this is an integration test I don't fully understand why we need explicit test for different data objects (tables,views,materialized views) - marquez doesn't recognize difference between trino data entities so there is no reason it would fail for particular one

Praveen2112 replied:
We need to ensure how the events are generated and linked if they are from different objects - like in case of `CREATE` do we map them to view or underlying table referenced table.

mgorsk1 replied:
ah you mean if we run CREATE VIEW VIEW AS SELECT * FROM TABLE TABLE_Y and then CREATE TABLE TABLE_X AS SELECT * FROM VIEW_A whether lineage info in inputs shows VIEW_A or TABLE_Y?

mgorsk1 replied:
if so - wouldn't it more fall into testing whatever fills out `QueryIOMetadata` object?

Praveen2112 replied:
Yeah - But we also need to ensure that - we use them in a correct way.

Praveen2112 replied:
We need to apply a filter on `directlyReferenced` to refer only actual tables instead of objects being referenced by views or MVs.

mgorsk1 replied:
but from my investigation:
- queryMetadata.getTables() won't show actual table if the view was used (and there is no reference chain present)
- queryIOMetadata.getInputs() will show tables referenced by views (and not list views at all)

so getMetadata().getTables() is not useful here because:
- if a view X of table A was used to create table B, then table A **won't be present in queryEventCompleted().getMetadata().getTables() at all** (the view will be though), but it will be present in queryCompletedEvent().getIoMetadata().getInputs() which is what we are already doing.

so how would I use directlyReferenced in such scenario?

Praveen2112 replied:
> if a view X of table A was used to create table B, then table A won't be present in queryEventCompleted().getMetadata().getTables() at all (the view will be though), but it will be present in queryCompletedEvent().getIoMetadata().getInputs() which is what we are already doing.

In case of populating the lineage - we need to map only the objects referred by a query right ? For a query like 
`CREATE TABLE B AS SELECT * FROM X`

 When we are looking at lineage events or the i/p required for `B` it should point only to `X` not `A` which is referred by `X` 

`queryMetadata.getTables()` should be even track the table `A` but the only difference is that `directlyReferenced` should be false in that case

mgorsk1 replied:
OpenLineage does not want to know about views, only tables. So if in your example B is a table, X is a view and A is a table (upon which view X is created), the lineage we want is:
- inputs: A
- outputs: B

this is by design

Praveen2112 replied:
In case `OpenLineage` we are tracking tables and columns and how are they linked, but in case of SQL views/MVs are also table like object right - so shouldn't we consider a view/MV as a part of lineage. For instance in case of Apache Atlas (a tool which tracks the lineage) we would treat View/MVs as a table so any PII or tags could be set on them as well.

mgorsk1 replied:
@mobuchowski could you chip in here? is there a place in OL where we reference views instead of tables? or - if a table is created based on a view should be store table as input and this view as a symlink?

Praveen2112 replied:
I'm not a OpenLineage expert but Table/View/MV are similar relation objects which we use in a query - though it could be resolved differently in the runtime - the query string which is being executed doesn't capture those abstraction - like `SELECT * FROM view_1` - Doesn't mention any details any information about `view_1` - it is being abstracted and the lineage should point to the abstraction instead of underlying references.

In case of Trino `SHOW TABLES` would treat everything as table object even if it is a view or a MVs.

mgorsk1 replied:
yes but imagine a scenario in which you do

- `CREATE VIEW VIEW_X AS SELECT * FROM TABLE_A`
- `CREATE TABLE TABLE_B AS SELECT * FROM VIEW_X`

in this scenario, if we have
- inputs: VIEW_X
- outputs: TABLE_B

then when someone removes VIEW_X from Trino, the lineage is gone as well - which is wrong. Even if VIEW_X is gone we should have lineage information present as actual flow of data was from TABLE_A.

Praveen2112 replied:
But this problem is irrespective of View/Table/MV right ? 

- `CREATE TABLE TABLE_X AS SELECT * FROM TABLE_A`
- `CREATE TABLE TABLE_B AS SELECT * FROM TABLE_X`

If someone drop `TABLE_X` then the lineage is cut off right `TABLE_B` would not be able to point to `TABLE_A` ? Ideally when we delete an entity it should be marked as inactive and the lineage should be maintained for historic purpose.

mobuchowski replied:
In ideal situation you'd get lineage from all parts: 
```
CREATE VIEW VIEW_X AS SELECT * FROM TABLE_A
```
would produce 
 - inputs: TABLE_A
 - outputs: VIEW_X

and as you said, 
```
CREATE TABLE TABLE_B AS SELECT * FROM VIEW_X
```
would produce 
 - inputs: VIEW_X
 - outputs: TABLE_B

> then when someone removes VIEW_X from Trino, the lineage would be gone as well - which is wrong. Even if VIEW_X is gone we should have lineage information present as actual flow of data was from TABLE_A.

I think this is implementation detail of data catalog (or other tool) that consumes OL events, I don't see why removal should result from immediate deletion of lineage?

mgorsk1 replied:
this is not the case for spark however, the views are translated into tables and tables are referenced in inputs. am I right?

mgorsk1 replied:
ok, I refactored the code so it:
- does not use queryCompletedEvent.getIOMetadata() at all
- uses queryCompletedEvent.getMetadata().getTables()
- inputs section is populated by entities found in getTables which have `isReferencedDirectly` set to `true`
- tables that are marked as not `isReferencedDirectly` go into symlinks.identifiers for corresponding table through referenceChain mapping

in other words:
1. on CREATE TABLE TABLE_B AS SELECT * FROM TABLE_A we should have:
- inputs: TABLE_A
- outputs: TABLE_B

2. on CREATE TABLE TABLE_C AS SELECT * FROM VIEW_A we should have:
- inputs: VIEW_A (plus symlinks.identifiers=TABLE_C)
- outputs: TABLE_B

let me know if that's what you both had in mind.

mgorsk1 replied:
```
  ~ curl localhost:32813/api/v1/namespaces/trino%3A%2F%2Ftrino-integration-test%3A1337/datasets/marquez.default.test_view | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1352  100  1352    0     0  10770      0 --:--:-- --:--:-- --:--:-- 10816
{
  "id": {
    "namespace": "trino://trino-integration-test:1337",
    "name": "marquez.default.test_view"
  },
  "type": "DB_TABLE",
  "name": "marquez.default.test_view",
  "physicalName": "marquez.default.test_view",
  "createdAt": "2024-04-24T13:47:33.895Z",
  "updatedAt": "2024-04-24T13:47:33.895Z",
  "namespace": "trino://trino-integration-test:1337",
  "sourceName": "default",
  "fields": [
    {
      "name": "nationkey",
      "type": null,
      "tags": [],
      "description": null
    },
    {
      "name": "regionkey",
      "type": null,
      "tags": [],
      "description": null
    },
    {
      "name": "name",
      "type": null,
      "tags": [],
      "description": null
    },
    {
      "name": "comment",
      "type": null,
      "tags": [],
      "description": null
    }
  ],
  "tags": [],
  "lastModifiedAt": null,
  "lastLifecycleState": "",
  "description": null,
  "currentVersion": "f9e2368e-6d1e-4b33-b6a2-2a61606f38ae",
  "columnLineage": null,
  "facets": {
    "schema": {
      "fields": [
        {
          "name": "nationkey"
        },
        {
          "name": "regionkey"
        },
        {
          "name": "name"
        },
        {
          "name": "comment"
        }
      ],
      "_producer": "https://github.com/trinodb/trino/plugin/trino-openlineage",
      "_schemaURL": "https://openlineage.io/spec/facets/1-1-0/SchemaDatasetFacet.json#/$defs/SchemaDatasetFacet"
    },
    "symlinks": {
      "_producer": "https://github.com/trinodb/trino/plugin/trino-openlineage",
      "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/SymlinksDatasetFacet.json#/$defs/SymlinksDatasetFacet",
      "identifiers": [
        {
          "name": "tpch.tiny.nation",
          "type": "TABLE",
          "namespace": "trino://trino-integration-test:1337"
        }
      ]
    }
  },
  "deleted": false
}
```

mobuchowski replied:
Looks great!

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,340 @@
+                client.emit(event);
+            }
+            catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
```
I think after all these changes we are no longer at risk for IllegalAccessException so just removed try catch and throws

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,340 @@
+
+        this.includeQueryTypes = listenerConfig.getIncludeQueryTypes();
+
+        this.queryStatisticsMapper = createQueryStatisticsObjectMapper();
```
is there anything particularly not ok with this approach? for JsonCodec I think I would also need to bring another dependency.

Praveen2112 replied:
`JsoncCodec` dependency is fine, it is from `io.airlift:json` - Plus `JsonCodec` unifies the list of modules required for `ObjectMapper` so we don't want to install them once again and it would be nice if we could have a unified version of Json ser/deser

mgorsk1 replied:
ok, I can add it but would need some help understanding this - mapJsonCodec enables me to serialize a class to json string or the other way around (it has methods fromJson and toJson) but that's not what I need, that's why i use directly object mapper - to get a property and put it in a hashmap (required by RunFacet.putAll())

alprusty commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,420 @@
+    public OpenLineageListener(OpenLineageClient client, OpenLineageListenerConfig listenerConfig)
+            throws Exception
+    {
+        this.client = requireNonNull(client);
```
Both Spark and Flink open lineage integration has made it configurable to name the `job.namespace` if an organization wants a custom name. Although there is [dataset naming conventions](https://openlineage.io/docs/spec/naming), there is no naming convention followed for `job.namespace` for different integrations.


I recently asked in the OL community channel as well
https://openlineage.slack.com/archives/C01CK9T7HKR/p1713900040070319

[Flink](https://github.com/OpenLineage/OpenLineage/blob/a97ae0bf33859855c33212d37735a2f520794eb4/integration/flink/app/src/main/java/io/openlineage/flink/OpenLineageFlinkJobListener.java#L55) override key:  `execution.job-listener.openlineage.namespace`

[Spark](https://github.com/OpenLineage/OpenLineage/blob/a97ae0bf33859855c33212d37735a2f520794eb4/integration/spark/app/src/main/java/io/openlineage/spark/agent/ArgumentParser.java#L39) override key: `spark.openlineage.namespace`

We should also give a similar override capability for Trino with `openlineage-event-listener.trino.namespace` property and can keep the default value as `trino://{host}:port`

mgorsk1 replied:
true, added `openlineage-event-listener.namespace` for overriding job namespace

mgorsk1 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,411 @@
+    {
+        RunFacet queryContextFacet = openLineage.newRunFacet();
+
+        java.util.Map<String, String> properties = new HashMap<String, String>();
```
done

mgorsk1 commented on plugin/trino-openlineage/pom.xml:
```diff
@@ -0,0 +1,184 @@
+            <groupId>com.fasterxml.jackson.datatype</groupId>
+            <artifactId>jackson-datatype-jsr310</artifactId>
+        </dependency>
+        <dependency>
```
done

mgorsk1 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/MarquezServer.java:
```diff
@@ -0,0 +1,130 @@
+        this.dockerContainerPostgres.start();
+        closer.register(this.dockerContainerPostgres::close);
+
+        this.dockerContainerAPI = new GenericContainer<>("marquezproject/marquez:" + version)
```
sure

mgorsk1 commented:
@Praveen2112 addressed latest remarks. As this went through quite a few iterations - do you think it's reached mergeable state? Subsequent changes related to transport generalization I would leave to @alprusty as he will be introducing kafka transport where such generalization will be useful.

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,424 @@
+                                        }
+
+                                        if (referenceDatasetName != null) {
+                                            directlyReferencedToSymlinksMap.get(referenceDatasetName).add(table);
```
What if we have an intermediate view ? i.e

CREATE TABLE table1;
CREATE VIEW view1 AS SELECT * FROM table1;
CREATE VIEW view2 AS SELECT * FROM view1;

CREATE TABLE new_table AS SELECT * FROM view2;

We could get a `view1` referenced but this map would throw a NPE.

mgorsk1 replied:
ok, so I propose to:
- catch this NPE and skip on error
- extend `include-query-type`s with `QueryType.DATA_DEFINITION`

in such case, when view is created we register this information, and then in case of the lineage scenario you brought up we will have 2 symlink entities (table and view) and they both will be already registered in OL

mgorsk1 replied:
ok, so as discussed - removed symlinks and rely on directly referenced entities from getTables only + included QueryType.DATA_DEFINITION so we capture view creation as well

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListenerModule.java:
```diff
@@ -0,0 +1,75 @@
+
+        install(conditionalModule(
+                OpenLineageListenerConfig.class,
+                config -> config.getTransport().equals(io.trino.plugin.openlineage.OpenLineageTransport.CONSOLE),
```
nit: static import directly

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListenerModule.java:
```diff
@@ -0,0 +1,75 @@
+        install(conditionalModule(
+                OpenLineageListenerConfig.class,
+                config -> config.getTransport().equals(io.trino.plugin.openlineage.OpenLineageTransport.CONSOLE),
+                conditionalBinder -> {
```
nit: It could be internalBinder and we use

                conditionalBinder -> conditionalBinder.bind(OpenLineageTransport.class).to(OpenLineageConsoleTransport.class)));

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListenerModule.java:
```diff
@@ -0,0 +1,75 @@
+
+        install(conditionalModule(
+                OpenLineageListenerConfig.class,
+                config -> config.getTransport().equals(io.trino.plugin.openlineage.OpenLineageTransport.HTTP),
```
static import and interalBinder

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/transport/OpenLineageTransport.java:
```diff
@@ -0,0 +1,22 @@
+
+import io.openlineage.client.transports.Transport;
+
+public interface OpenLineageTransport
```
❤️

Praveen2112 replied:
This looks even better

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/transport/OpenLineageHttpTransport.java:
```diff
@@ -0,0 +1,69 @@
+    @Inject
+    public OpenLineageHttpTransport(OpenLineageClientHttpTransportConfig config)
+    {
+        this.config = config;
```
Can we avoid maintaining config here - In general `Config` objects are mutable and they could potentially change in the future - So we could directly create some field for url and endpoint and update them accordingly or even initialize `HttpConfig` in the ctor.

mgorsk1 replied:
sure, created explicit attributes for this class

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/transport/OpenLineageHttpTransport.java:
```diff
@@ -0,0 +1,69 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.openlineage.transport;
```
Since this is specific to `http` can we move to a package like `io.trino.plugin.openlineage.transport.http` and for Config object also so for kafka also we could have a dedicated package

Praveen2112 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/TestMarquezIntegration.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Test
+    void testCreateTableAsSelectFromView()
```
Can we check some sort of nested view - like the one we discussed above.

mgorsk1 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/TestMarquezIntegration.java:
```diff
@@ -0,0 +1,145 @@
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestMarquezIntegration
```
how about `TestOpenLineageEventListenerMarquezIntegration`?

Praveen2112 commented:
Can we squash all the commits into 1?

mgorsk1 commented:
@Praveen2112 sure, done!

mosabua commented:
Discussed in the Trino Contributor Call - https://github.com/trinodb/trino/wiki/Contributor-meetings#trino-contributor-call-23-may-2024

Seems like this is ready to go and just waiting for @Praveen2112 to merge.

Praveen2112 commented on plugin/trino-openlineage/pom.xml:
```diff
@@ -0,0 +1,180 @@
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
```
Can we remove this new line ?

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,362 @@
+    @Inject
+    public OpenLineageListener(OpenLineageClient client, OpenLineageListenerConfig listenerConfig)
+    {
+        this.client = requireNonNull(client);
```
nit: It would be nice if we could have message associated with requireNonNull

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,362 @@
+                this.includeQueryTypes));
+    }
+
+    private Boolean queryTypeSupported(QueryContext queryContext)
```
nit: It can be primitive `boolean`

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,362 @@
+
+    private String getDatasetName(TableInfo tableInfo)
+    {
+        return format("%s.%s.%s", tableInfo.getCatalog(), tableInfo.getSchema(), tableInfo.getTable());
```
nit: Instead we could try like
```
return getDatasetName(tableInfo.getCatalog(), tableInfor.getSchema(), tableInfo.getTable())

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java:
```diff
@@ -0,0 +1,362 @@
+        return format("%s.%s.%s", catalogName, schemaName, tableName);
+    }
+
+    private static String resolveEntityType(TableInfo tableInfo)
+    {
+        if (tableInfo.getViewText().isPresent()) {
+            return "VIEW";
+        }
+        else if (!tableInfo.getFilters().isEmpty()) {
+            return "FILTER";
+        }
+        else {
+            return "TABLE";
+        }
+    }
```
This method is unused right ?

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageListenerConfig.java:
```diff
@@ -0,0 +1,134 @@
+
+    @Config("openlineage-event-listener.disabled-facets")
+    @ConfigDescription("Which facets should be removed from OpenLineage events.")
+    public OpenLineageListenerConfig setDisabledFacets(String disabledFacets)
```
I think Airlift has a util methods which allows us to directly inject the enum and we could inject them as `Set<OpenLineageTrinoFacet>` and same for QueryTypes as well. 

Can we add test for `OpenLineageListenerConfig`

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/OpenLineageListenerConfig.java:
```diff
@@ -0,0 +1,134 @@
+        return this;
+    }
+
+    public Optional<String> getNamespace()
```
Optional<@NotEmpty String> to ensure we don't pass empty namesapce

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/http/OpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,139 @@
+
+import static java.lang.String.format;
+
+public class OpenLineageClientHttpTransportConfig
```
Can we have some test for this ` OpenLineageClientHttpTransportConfig` - We do have simple tests for config objects like `TestKafkaConfig` which tests the default value and runtime values.

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/http/OpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,139 @@
+        try {
+            this.headers = headers
+                    .stream()
+                    .collect(Collectors.toMap(kvs -> kvs.split(":", 2)[0], kvs -> kvs.split(":", 2)[1]));
```
keyValue ?

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/config/http/OpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,139 @@
+    private String endpoint;
+    private Optional<String> apiKey = Optional.empty();
+    private Duration timeout = new Duration(5000, TimeUnit.MILLISECONDS);
+    private Map<String, String> headers = new HashMap<>();
+    private Map<String, String> urlParams = new HashMap<>();
```
ImmutableMap#of

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/transport/http/OpenLineageHttpTransport.java:
```diff
@@ -0,0 +1,82 @@
+        this.endpoint = config.getEndpoint();
+        this.timeout = (int) config.getTimeout().toMillis();
+        this.apiKey = config.getApiKey();
+        this.urlParams = config.getUrlParams();
+        this.headers = config.getHeaders();
```
Can we use `ImmutableMap#copyOf`

Praveen2112 commented on plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/transport/http/OpenLineageHttpTransport.java:
```diff
@@ -0,0 +1,82 @@
+        this.url = config.getUrl();
+        this.endpoint = config.getEndpoint();
+        this.timeout = (int) config.getTimeout().toMillis();
+        this.apiKey = config.getApiKey();
```
Can we have a singleton instance of `ApiKeyTokenProvider` instead of maintaining the `String`

Praveen2112 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/MarquezServer.java:
```diff
@@ -0,0 +1,171 @@
+        this(DEFAULT_VERSION, false);
+    }
+
+    public MarquezServer(Boolean ui)
```
Can we use primitive `boolean` and uiEnabled for variableName.

Praveen2112 replied:
Do we need this toggle ? Can we enable it out by default ?

mosabua reviewed (approved):
From docs perspective only, this looks good enough to merge. I plan on updating and expanding docs in follow up PR. I also expect follow up code PRs to come.

Praveen2112 commented on plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/TestOpenLineageClientHttpTransportConfig.java:
```diff
@@ -0,0 +1,65 @@
+                .put("openlineage-event-listener.transport.timeout", "30s")
+                .put("openlineage-event-listener.transport.headers", "header1:value1,header2:value2")
+                .put("openlineage-event-listener.transport.url-params", "urlParam1:urlVal1,urlParam2:urlVal2")
+
```
We don't need a new line here.

Praveen2112 reviewed (approved):
LGTM. Can we add @takezoe as a co-author as a part of the commit message.

Praveen2112 commented:
Thanks a lot for working on this.

alprusty commented:
🎉

takezoe commented:
🎉

mgorsk1 commented:
🕺🎉

nineinchnick commented:
@colebow this is included in 449 but its missing release notes

ebyhr commented:
@nineinchnick We added it in https://github.com/trinodb/trino/pull/22224

Chaho12 commented:
@mgorsk1 Thx for such feature. I was wondering if there was any plan to add filter on table, so that it doesn't collect for certain tables/schemas/catalog.

mgorsk1 commented:
Hey @Chaho12 happy to help! We have not had such need but it seems like a valid extension indeed so feel
free to submit feature request and maybe even PR, I am sure community will be happy to accept it.
